### PR TITLE
Added shortest_paths algorithm API and example

### PR DESCRIPTION
**Pull request by bysheva**
# Description
Please include a summary of the change. For large or complex changes please include enough information to introduce your change and explain motivation for it.

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

File: cpp/oneapi/dal/algo/shortest_paths.hpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths.hpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel.hpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel_cpu.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/backend/cpu/traverse_default_kernel_cpu.cpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/common.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/common.cpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/common.hpp [hunk 5]
```diff
@@ -0,0 +1,208 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/common.hpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/detail/select_kernel.hpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/detail/select_kernel.hpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/detail/traverse_default_kernel.cpp [hunk 7]
```diff
@@ -0,0 +1,23 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/detail/traverse_default_kernel.cpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/detail/traverse_default_kernel.hpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/detail/traverse_default_kernel.hpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/detail/traverse_ops.hpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/detail/traverse_ops.hpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/detail/traverse_types.hpp [hunk 10]
```diff
@@ -0,0 +1,37 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/detail/traverse_types.hpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/detail/traverse_types.hpp [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+namespace oneapi::dal::preview::shortest_paths::detail {
+
+class traverse_result_impl;
+
+template <typename Graph, typename Task>
+class traverse_input_impl : public base {
+public:
+    traverse_input_impl(const Graph& g) : graph_data(g) {}
+
+    const Graph& graph_data;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/detail/traverse_types.hpp:34]: While it also present in another algs, I suggest to use encapsulation.

File: cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp [hunk 12]
```diff
@@ -0,0 +1,25 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/test/shortest_paths_test.cpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/traverse.hpp [hunk 13]
```diff
@@ -0,0 +1,30 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/traverse.hpp:2]: 2021

File: cpp/oneapi/dal/algo/shortest_paths/traverse_types.cpp [hunk 14]
```diff
@@ -0,0 +1,54 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/traverse_types.cpp:2]: 2021

File: README.md [hunk 15]
```diff
@@ -238,8 +238,10 @@ In C++ APIs, technical preview features are located in `daal::preview` and `onea
 The preview features list:
 - Graph Analytics:
 	- Undirected graph without edge and vertex weights (`undirected_adjacency_vector_graph`), where vertex indices can only be of type int32
+  - Directed graph with and without edge weights (`directed_adjacency_vector_graph`), where vertex indices can only be of type int32, edge weights can be int32 and double
```
Comment (outoftardis) [inline README.md:241]:
```suggestion
  - Directed graph with and without edge weights (`directed_adjacency_vector_graph`), where vertex indices can only be of type int32, edge weights can be of type int32 or double
```

File: cpp/oneapi/dal/algo/shortest_paths/traverse_types.cpp [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+*******************************************************************************/
+
+#include "oneapi/dal/algo/shortest_paths/traverse_types.hpp"
+
+namespace oneapi::dal::preview::shortest_paths {
+
+class detail::traverse_result_impl : public base {
+public:
+    table distances;
+    table predecessors;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/traverse_types.cpp:24]: Encapsulation

File: cpp/oneapi/dal/algo/shortest_paths/traverse_types.hpp [hunk 17]
```diff
@@ -0,0 +1,104 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/shortest_paths/traverse_types.hpp:2]: 2021

File: cpp/oneapi/dal/graph/detail/directed_adjacency_vector_graph_topology_builder.hpp [hunk 18]
```diff
@@ -0,0 +1,52 @@
+/* file: directed_adjacency_vector_graph_topology_builder.hpp */
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/graph/detail/directed_adjacency_vector_graph_topology_builder.hpp:3]: 2021

File: cpp/oneapi/dal/traverse.hpp [hunk 19]
```diff
@@ -0,0 +1,33 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline cpp/oneapi/dal/traverse.hpp:2]: 2021

File: examples/oneapi/cpp/source/shortest_paths/shortest_paths_batch.cpp [hunk 20]
```diff
@@ -0,0 +1,59 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
```
Comment (averbukh) [inline examples/oneapi/cpp/source/shortest_paths/shortest_paths_batch.cpp:2]: 2021

File: examples/oneapi/cpp/source/shortest_paths/shortest_paths_batch.cpp [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+#include "oneapi/dal/io/graph_csv_data_source.hpp"
+#include "oneapi/dal/io/load_graph.hpp"
+
+namespace dal = oneapi::dal;
+using namespace dal::preview::shortest_paths;
+
+int main(int argc, char** argv) {
+    const auto filename = get_data_path("weighted_edge_list.csv");
+
+    //read the graph
```
Comment (averbukh) [inline examples/oneapi/cpp/source/shortest_paths/shortest_paths_batch.cpp:31]: Add space between // and 'read'

File: cpp/oneapi/dal/algo/shortest_paths/traverse_types.hpp [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/// @file
+/// Contains the definition of the input and output for Shortest Paths
```
Comment (outoftardis) [inline cpp/oneapi/dal/algo/shortest_paths/traverse_types.hpp:18]:
```suggestion
/// Contains the definition of the input and output for the Shortest Paths
```

**Review by orrrrtem — APPROVED** [review]
To be continued...

**Pull request merged**